### PR TITLE
[pipeline] Provide Parallel Sort Operator

### PR DESCRIPTION
**Pull request by Pslydhh**
This PR is used to provide data parallelism for Sort.
Before this PR:
- We append LocalExchange to all pipelines to gather all chunks in one sort operator.

After this PR:

- We append PartitionSortOperator to every pipeline to sort chunks, this provdie data parallelism.
- We merge all sorted chunks from all PartitionSortOperators through LocalMergeSortOperator, like It in ExchangeMergeSortOperator.

https://github.com/StarRocks/starrocks/issues/1267

### CODE REVIEW

File: be/src/exec/pipeline/sort/sort_context.h [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+        // for data range from one DataSegment, used to collect chunks.
+        std::vector<uint32_t> selective_values;
+        selective_values.reserve(needed_rows);
+        auto min_heap_entry = _data_segment_heaps[0];
+        ChunkPtr result_chunk = min_heap_entry->chunk->clone_empty_with_slot(needed_rows);
+
+        // Optimization for single thread.
+        DCHECK(_data_segment_heaps.size() == 1);
+        if (_data_segment_heaps.size() == 1) {
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/sort/sort_context.h:88]: Maybe we should remove this `DCHECK`?

File: be/src/exec/pipeline/sort/sort_context.h [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+public:
+    explicit SortContext(int64_t limit, const int32_t num_right_sinkers, const std::vector<bool>& is_asc_order,
+                         const std::vector<bool>& is_null_first)
+            : _limit(limit), _num_partition_sinkers(num_right_sinkers), _comparer(limit, is_asc_order, is_null_first) {
+        _chunks_sorter_partions.reserve(num_right_sinkers);
+        _data_segment_heaps.reserve(num_right_sinkers);
+    }
+
+    void finish_partition(uint64_t partition_rows) {
+        _total_rows += partition_rows;
```
[the remark this answers is not in the document]
Comment (Pslydhh) [reply be/src/exec/pipeline/sort/sort_context.h:33]: Yes, it has data race.

File: be/src/exec/pipeline/sort/partition_sort_sink_operator.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    // materialize all sorting columns: replace old columns with evaluated columns
+    const size_t row_num = chunk->num_rows();
+    const auto& slots_in_row_descriptor = _materialized_tuple_desc->slots();
+    const auto& slots_in_sort_exprs = _sort_exec_exprs.sort_tuple_slot_expr_ctxs();
+
+    DCHECK_EQ(slots_in_row_descriptor.size(), slots_in_sort_exprs.size());
+
+    for (size_t i = 0; i < slots_in_sort_exprs.size(); ++i) {
+        ExprContext* expr_ctx = slots_in_sort_exprs[i];
+        ColumnPtr col = expr_ctx->evaluate(chunk);
```
Comment (kangkaisen) [inline be/src/exec/pipeline/sort/partition_sort_sink_operator.cpp:58]: These codes are repeated with `SortSinkOperator::_materialize_chunk_before_sort` and `TopNNode::_materialize_chunk_before_sort`

File: be/src/exec/pipeline/sort/local_merge_sort_source_operator.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "column/vectorized_fwd.h"
+#include "exec/pipeline/source_operator.h"
+#include "exec/sort_exec_exprs.h"
+
+namespace starrocks {
+namespace pipeline {
+class SortContext;
+
+class LocalMergeSortSourceOperator final : public SourceOperator {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/sort/local_merge_sort_source_operator.h:13]: Add some comment for this class

File: be/src/exec/vectorized/topn_node.cpp [hunk 5]
```diff
@@ -220,39 +222,25 @@ ChunkPtr TopNNode::_materialize_chunk_before_sort(Chunk* chunk) {
...
+    auto degree_of_parallelism =
+            down_cast<SourceOperatorFactory*>(operators_sink_with_sort[0].get())->degree_of_parallelism();
+    auto sort_context = std::make_shared<SortContext>(_limit, degree_of_parallelism, _is_asc_order, _is_null_first);
 
-    // add SortSinkOperator to this pipeline
-    auto sort_sink_operator = std::make_shared<SortSinkOperatorFactory>(
-            context->next_operator_id(), id(), chunks_sorter, _sort_exec_exprs, _order_by_types,
-            _materialized_tuple_desc, child(0)->row_desc(), _row_descriptor);
-    operators_sink_with_sort.emplace_back(std::move(sort_sink_operator));
+    auto partition_sort_sink_operator = std::make_shared<PartitionSortSinkOperatorFactory>(
```
Comment (kangkaisen) [inline be/src/exec/vectorized/topn_node.cpp:231]: Do still need `SortSinkOperatorFactory` and `SortSourceOperatorFactory`?

**Review by satanson — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

File: be/src/exec/vectorized/topn_node.cpp (hunk 5 above)
Comment (Pslydhh) [reply be/src/exec/vectorized/topn_node.cpp:231]: We will remove them.

File: be/src/exec/pipeline/sort/partition_sort_sink_operator.cpp (hunk 3 above)
Comment (Pslydhh) [reply be/src/exec/pipeline/sort/partition_sort_sink_operator.cpp]: Will use a common materialize_chunk_before_sort in ChunksSorter

**Review by satanson — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**